### PR TITLE
Sd kl duplicate items

### LINKED ISSUE #6: 6. As a user, I want to be alerted when I’m entering an item that is the same as something already on my list so I can avoid duplicate items.
AC:

* Show an error message if the user tries to submit a new item that has the exact same name as an existing item
* Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalization has been normalized and punctuation has been removed
* The user’s original input should be what gets saved in the database

### PR DESCRIPTION
**Pull request by stef-codes**
## Description

This code modifies the `<AddItem />` component to show error messages for duplicate items. There is data clean up for punctuation and capitalization to ensure correct data comparison. The approach was to use regex to remove punctuation and the `toLowerCase()` built in method for capitalization. This issue was informative related to implementing aspects of data quality. 

## Related Issue

closes issue #6 

## Acceptance Criteria

- [x] Show an error message if the user tries to submit a new item that has the exact same name as an existing item
- [x] Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalization has been normalized and punctuation has been removed
- [x] The user’s original input should be what gets saved in the database

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
| ✓ | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

### Before
collab_lab_issue_6_before

### After
collab_lab_issue_6_after

## Testing Steps / QA Criteria

- From your terminal, run git checkout `sd-kl-duplicate-items`.
- Once on the `sd-kl-duplicate-items` branch, run npm start to load the browser.
- Visit the `<List/>` component at http://localhost:3000/list.
- See the items in your list.
- Click `Add Item` and add a new item to your list with punctuation or capitalization. 
- Visit the `<List/>` component at http://localhost:3000/list and you should see your item.
- Add the same item to your list again with or without punctuation or capitalization.
- You should receive an error message.

### CODE REVIEW

**Comment by stef-codes** [comment]
Adding before and after images

File: src/components/AddItem/AddItem.component.jsx [hunk 1]
```diff
@@ -27,24 +28,35 @@ const AddItem = () => {
...
-        lastPurchaseDate: lastPurchaseDate,
-      });
 
-    // Conditionally renders tooltip after adding our item to firebase.
-    setIsAdded(true);
-    setTimeout(() => {
-      setIsAdded(false);
-    }, 1200);
+    // Clean Input
+    const cleanInput = itemName.toLowerCase().replace(/[^\w\s]|/g, '');
```
Comment (jendevelops) [inline src/components/AddItem/AddItem.component.jsx]: Nice and clean!

File: src/components/AddItem/AddItem.component.jsx [hunk 2]
```diff
@@ -27,24 +28,35 @@ const AddItem = () => {
...
+            resupplyPeriod: resupplyPeriod,
+            id: itemId,
+            lastPurchaseDate: lastPurchaseDate,
+          });
+          setIsAdded(true);
+          setTimeout(() => {
+            setIsAdded(false);
+          }, 1200);
+        } else {
+          alert('already exists');
```
Comment (jendevelops) [inline src/components/AddItem/AddItem.component.jsx]: Javascript alert messages are not the most accessible option when displaying an error.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/forms/alerts#Checking_for_validity_and_notifying_the_user_2

Maybe displaying errors is something we can discuss during the Weekly Sync about how to approach it in a different way. @joeylaguna

File: src/components/Home/Home.component.jsx [hunk 3]
```diff
@@ -1,7 +1,48 @@
...
+import React, { useState } from 'react';
+import getToken from '../../lib/tokens';
+import { useHistory } from 'react-router-dom';
 const Home = () => {
-  return <div> Home Page</div>;
+  const [localToken, setLocalToken] = useState(localStorage.getItem('token'));
+  const history = useHistory();
+
+  const generateToken = () => {
+    console.log('called generate token');
```
Comment (jendevelops) [inline src/components/Home/Home.component.jsx]: Does this still need to be in the code?

File: src/components/Home/Home.component.jsx [hunk 4]
```diff
@@ -1,7 +1,48 @@
...
+  // Redirect to the Add Item View
+  const redirectAddItem = () => {
+    history.push('/addItem');
+  };
+  return (
+    <div>
+      {localToken ? (
+        <>
+          <h1 className="page__title">Welcome Back!</h1>
+          <br /> <br />
```
Comment (jendevelops) [inline src/components/Home/Home.component.jsx:57]: We'll eventually want to get rid of these when more CSS styling is implemented

**Review by jendevelops — APPROVED** [review]
Really great job! I just had a few comments here and there, but none are blockers!

File: PROJECT-BRIEF.md [hunk 5]
```diff
@@ -142,7 +142,7 @@ When you and your pair partner have working code that you believe is ready to be
 3. Incorporate feedback from the other pair team into your work until both you and they are satisfied the code is ready to be merged.
 4. Request that one of the mentors review the PR for final approval.
 5. Once approved, merge the PR into `main`. (Your code will be built and deployed to production automatically thanks to [Netlify](https://www.netlify.com/).)
-6. Check your work on the [production site]({link to cohort live site}).
+6. Check your work on the [production site](https://tcl-11-smart-shopping-list.netlify.app/).
```
Comment (joeylaguna) [inline PROJECT-BRIEF.md]: Thanks for adding this link 😄

File: src/components/AddItem/AddItem.component.jsx [hunk 6]
```diff
@@ -27,24 +28,35 @@ const AddItem = () => {
...
+    const cleanInput = itemName.toLowerCase().replace(/[^\w\s]|/g, '');
 
-    setItemName('');
+    db.collection(collectionTokenName)
+      .get()
+      .then(snapshot => {
+        const items = snapshot.docs
+          .map(query => query.data())
+          .map(data => data.name.toLowerCase().replace(/[^\w\s]|/g, ''));
+        console.log(items);
```
Comment (joeylaguna) [inline src/components/AddItem/AddItem.component.jsx]: Not sure if this was used for testing, but if it was these two console.logs can be removed

**Review by daredtech — APPROVED** [review]
Hi team.
Acceptance criteria have been met. 

Loved the use of regex and appreciate the comments. Moreover, you did a great job figuring the firestore parsing out!

However, I could not get `Success! Item added.`  displayed. Line 74, in `AddItem`. 
- Also, I was wondering, what is your opinion on space `' '` ? 
For instance, one could save two the same values with  `' '`. See documents `KqynytV0ZnnlXexVvXpX` and `PH2QkmABj0rzoUZMr1DH` under `curd wast take`.
- And should 'hot dog' and 'hotdog' be the same or separate items?

Looking forward to hearing your perspective on that. Thanks.

**Comment by stef-codes** [comment]
> Hi team.
> Acceptance criteria have been met.
> 
> Loved the use of regex and appreciate the comments. Moreover, you did a great job figuring the firestore parsing out!
> 
> However, I could not get `Success! Item added.` displayed. Line 74, in `AddItem`.
> 
> * Also, I was wondering, what is your opinion on space `' '` ?
>   For instance, one could save two the same values with  `' '`. See documents `KqynytV0ZnnlXexVvXpX` and `PH2QkmABj0rzoUZMr1DH` under `curd wast take`.
> * And should 'hot dog' and 'hotdog' be the same or separate items?
> 
> Looking forward to hearing your perspective on that. Thanks.

Thanks for the comment! It does make sense that spaces should be taken into account. The code is updated with that change. Thanks for mentioning that extra feature 🚀

**Comment by brewswain** [comment]
Hey everyone. The AC is fit perfectly! 
From a code point of view, I really liked the regex implementation, everything works as expected.

**Review by brewswain — APPROVED** [review]

**Comment by daredtech** [comment]
@CodeMeKathy @stef-codes 

Is there a reason for using 

```
import 'firebase/firestore';
import * as firebase from 'firebase/app';
```

instead of 

`import * as firebase from '../Firebase/Firebase.component';`
?

Why do another import, and then declare `const db = firebase.firestore();` if `firebase.dataBase` from `import * as firebase from '../Firebase/Firebase.component';` does exactly the same thing? Please let me know.

**Comment by joeylaguna** [comment]
@CodeMeKathy it looks like there are some merge conflicts on this branch. Let me know if you want help resolving them!

**Comment by joeylaguna** [comment]
Updated this branch with the latest on `main` and removed an unused import that was causing this build to fail. This is good to merge now :)

**Pull request merged**